src/Common/OpacityTable.cpp

```diff
@@ -0,0 +1,308 @@
+          ++i;
+          j = 0;
+        }
+
+      }
+      else {
+        ExceptionHandler::getIstance().raise("Error: The specified opacity file cannot be read!");
+      }
+
+    }
+    file.close();
+  }
+
+  // Set fcol2
+  if (lombardi) {
+    fcol2 = fcol * fcol * 4.0 * pi;
+  }
+  else {
+    fcol2 = fcol * fcol;
+  }
+}
+
+
+
+//=================================================================================================
+//  OpacityTable::~OpacityTable()
+/// OpacityTable class destructor
+//=================================================================================================
+template <int ndim>
+OpacityTable<ndim>::~OpacityTable()
+{
+  for (int i = 0; i < ndens; i++) {
+    delete[] eos_energy[i];
+    delete[] eos_mu[i];
+    delete[] kappa_table[i];
+    delete[] kappar_table[i];
+    delete[] kappap_table[i];
+    delete[] eos_gamma[i];
+  }
+
+  delete[] eos_energy;
+  delete[] eos_mu;
+  delete[] kappa_table;
+  delete[] kappar_table;
+  delete[] kappap_table;
+  delete[] eos_gamma;
+}
+
+
+// ------------------------------------------------------------------------------------------//
+//			find  closest index in list for value: level
+// ------------------------------------------------------------------------------------------//
+template <typename BidirectionalIterator, typename T>
+BidirectionalIterator getClosest
+ (BidirectionalIterator first,
+  BidirectionalIterator last,
+  const T &value)
+{
+  BidirectionalIterator before = std::lower_bound(first, last, value);
+
+  if (before == first) return first;
+  if (before == last)  return --last;
+
+  BidirectionalIterator after = before;
+  --before;
+
+  return (*after - value) < (value - *before) ? after : before;
+}
+
+
+template <typename BidirectionalIterator, typename T>
+std::size_t getClosestIndex(BidirectionalIterator first,
+                            BidirectionalIterator last,
+                            const T &value)
+{
+  return std::distance(first, getClosest(first, last, value));
+}
+
+
+
+//=================================================================================================
+//  OpacityTable::GetIDens()
+/// GetIDens returns table index for density
+//=================================================================================================
+template <int ndim>
+int OpacityTable<ndim>::GetIDens
+ (const FLOAT rho)
+{
+  return getClosestIndex(eos_dens, eos_dens + ndens, rho);
+}
+
+
+
+//=================================================================================================
+//  OpacityTable::GetITemp()
+/// GetITemp returns table index for temperature
+//=================================================================================================
+template <int ndim>
+int OpacityTable<ndim>::GetITemp
+ (const FLOAT temp)
+{
+  return getClosestIndex(eos_temp, eos_temp + ntemp , temp);
+}
+
+
+
+//=================================================================================================
```

Comment:
Can we have and use an overloaded version of this function that takes iDens as an int?